### PR TITLE
Merklize accounts internal state

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem
We need a way to generate proofs of account state changes

#### Summary of Changes
Replace hash of concatenated accounts state hashes with a Merkle root

### CODE REVIEW

**Comment by t-nelson** [comment]
@mvines I *think* this is the only place where SPV and verifying accounts loaded from snapshots overlap.  Should be good to go on cutting someone loose on the latter

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -443,11 +444,9 @@ impl Accounts {
         if account_hashes.is_empty() {
             None
         } else {
-            let mut hasher = Hasher::default();
-            for (_, _, hash) in account_hashes {
-                hasher.hash(hash.as_ref());
-            }
-            Some(hasher.result())
+            let leaf_data: Vec<_> = account_hashes.iter().map(|ah| ah.2.as_ref()).collect();
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: if you rebase, this gets prettier, I think ;)

Comment (t-nelson) [reply runtime/src/accounts.rs]: Sure does!  Just resolved the conflict :)

**Comment by sakridge** [comment]
Speed difference?

**Comment by t-nelson** [comment]
@sakridge `runtime/benches/accounts.rs` is my friend for that info?

**Comment by sakridge** [comment]
@t-nelson I don't think there exists a unit test for this at the moment.

**Comment by t-nelson** [comment]
@sakridge Yeah saw that.  Adding one now.

**Comment by sakridge** [comment]
@t-nelson You might check this one out:
https://github.com/solana-labs/solana/pull/5075
run with `cd banking_bench; RUST_LOG=banking_bench=info cargo run --bin solana-banking-bench --release`

There should be a print of "average: xx us tx_average: xx us" average contains the time with the accounts hash_internal_state.

**Comment by t-nelson** [comment]
@sakridge I used the "test" crate, as was already in place for some benches under runtime.  https://github.com/t-nelson/solana/commit/22fc26886c63e6f968b7b84ce7ceea3193b2cc0d#diff-7348e822a67e92dad03464ab694a595eR61.  No problem to write up something more thorough if it's insufficient

It's takes about twice as long

Hash of concatenated account hashes (old):
`test test_accounts_hash_internal_state ... bench:     285,918 ns/iter (+/- 28,519)`

Merklized account hashes (new):
`test test_accounts_hash_internal_state ... bench:     584,399 ns/iter (+/- 255,462)`

**Comment by sakridge** [comment]
@t-nelson Yea I ran it on the benchmark I referenced, it is about 15% reduction in performance just from this. That's pretty huge. Can we make this any faster?

**Comment by t-nelson** [comment]
Assuming we don't need the actual account hash later (they appear to be discarded now), I think I might be able to tweak it to pass the serialization of the account state directly to the Merkle Tree ctor.  This would eliminate N hashing operations on something that is doing ~3N now

**Comment by t-nelson** [comment]
@sakridge is the above bench something we want in tree?  I can PR it separately if so

I'll look into reducing hashing operations later today

**Comment by t-nelson** [comment]
@sakridge ~20% improvement with the proposed reduction of hashes

Without proposal:
`test test_accounts_hash_internal_state ... bench:     578,833 ns/iter (+/- 57,388)`

With proposal:
`test test_accounts_hash_internal_state ... bench:     471,129 ns/iter (+/- 78,149)`

**Comment by sakridge** [comment]
@t-nelson yea we should add the bench to the tree. separate PR is great.

**Comment by garious** [comment]
@t-nelson, what's going on with this PR?

**Comment by t-nelson** [comment]
@garious I need to reduce negative effects on perf

**Comment by aeyakovenko** [comment]
@t-nelson @sakridge close this one?

**Comment by t-nelson** [comment]
Superseded by https://github.com/solana-labs/solana/pull/5573

**Pull request closed without merging**